### PR TITLE
Fixing: Torch 2.0 fails with cuda=False

### LINKED ISSUE #280: Upgrade to torch 2.0
[Pytorch 2.0](https://pytorch.org/get-started/pytorch-2.0/) was released a few weeks ago and is included as the default for Google Colab notebooks now.

**Ideal Result**: If CTGAN works on torch 2.0 (or if we can make minor modifications for it to work), then we should consider _uncapping_ the max version of `torch`

Alternatively, we can just increase the cap for now to include 2.0.

### LINKED ISSUE #288: Torch 2.0 fails with cuda=False
### Environment Details

Please indicate the following details about the environment in which you found the bug:

* CTGAN version: Master
* Python version: Any
* Operating System: Mac, Linux
* Pytorch version: 2.0.0

### Error Description

With Torch 2.0, the demo code fails whenever the `cuda` parameter is set to False. This is a problem because it is False by default in Linux.

### Steps to reproduce

```python
data = load_demo()
discrete_columns = [
    'workclass',
    'education',
    'marital-status',
    'occupation',
    'relationship',
    'race',
    'sex',
    'native-country',
    'income'
]

ctgan = CTGAN(epochs=10, cuda=False)
ctgan.fit(data, discrete_columns)

# Synthetic copy
samples = ctgan.sample(1000)
```

### PR DESCRIPTION
**Pull request by amontanez24**
Resolves #288 
Resolves #280

### CODE REVIEW

File: ctgan/synthesizers/ctgan.py [hunk 1]
```diff
@@ -197,15 +196,12 @@ def _gumbel_softmax(logits, tau=1, hard=False, eps=1e-10, dim=-1):
         Returns:
             Sampled tensor of same shape as logits from the Gumbel-Softmax distribution.
         """
-        if version.parse(torch.__version__) < version.parse('1.2.0'):
```
Comment (amontanez24) [inline ctgan/synthesizers/ctgan.py:200]: Once updating to torch >2.0, this problem seems. to occur again. Although not too frequently, `gumbel_softmax` sometimes returns nans.

File: ctgan/synthesizers/ctgan.py [hunk 2]
```diff
@@ -381,7 +377,7 @@ def fit(self, train_data, discrete_columns=(), epochs=None):
                         real_cat, fake_cat, self._device, self.pac)
                     loss_d = -(torch.mean(y_real) - torch.mean(y_fake))
 
-                    optimizerD.zero_grad()
+                    optimizerD.zero_grad(set_to_none=False)
```
Comment (amontanez24) [inline ctgan/synthesizers/ctgan.py:380]: the default was changed in torch>=2

File: ctgan/synthesizers/ctgan.py [hunk 3]
```diff
@@ -197,15 +196,12 @@ def _gumbel_softmax(logits, tau=1, hard=False, eps=1e-10, dim=-1):
         Returns:
             Sampled tensor of same shape as logits from the Gumbel-Softmax distribution.
         """
-        if version.parse(torch.__version__) < version.parse('1.2.0'):
-            for i in range(10):
-                transformed = functional.gumbel_softmax(logits, tau=tau, hard=hard,
```
Comment (amontanez24) [inline ctgan/synthesizers/ctgan.py:202]: @fealho Do you know the importance of using this function? It seems like it is a legacy function according to https://pytorch.org/docs/stable/generated/torch.nn.functional.gumbel_softmax.html

Comment (fealho) [reply ctgan/synthesizers/ctgan.py:202]: It's because of issue #73. It's been almost 3 years though, so it should be alright to drop support.

**Review by fealho — APPROVED** [review]

**Review by frances-h — APPROVED** [review]

**Pull request merged**